🎉 Source Retently: Add campaigns, feedback, outbox, templates streams

henriblancke opened the pull request:
## What
This PR pulls in more streams for retently, including:
* campaigns
* feedback
* outbox
* templates

## How
Added streams for campaigns, feedback, outbox and templates endpoints.

### Other improvements
* Increase the request limit to speed up data retrieval
* Fix customers json schema (properties -> value can contain integers). This change breaks the backward compatibility acceptance test.
* Updates to spec file to make acceptance tests pass

## Pre-merge Checklist
Expand the relevant checklist and delete the others.

<details><summary><strong>New Connector</strong></summary>

### Community member or Airbyter

- [ ] **Community member?** Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [ ] Secrets in the connector's spec are annotated with `airbyte_secret`
- [ ] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [ ] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [ ] `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
    - [ ] `docs/integrations/README.md`
    - [ ] `airbyte-integrations/builds.md`
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)
- [ ] After the connector is published, connector added to connector index as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)
- [ ] Seed specs have been re-generated by building the platform and committing the changes to the seed spec files, as described [here](https://docs.airbyte.io/connector-development#publishing-a-connector)

</details>

<details><summary><strong>Updating a connector</strong></summary>

### Community member or Airbyter

- [x] Grant edit access to maintainers ([instructions](https://docs.github.com/en/github/collaborating-with-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork#enabling-repository-maintainer-permissions-on-existing-pull-requests))
- [x] Secrets in the connector's spec are annotated with `airbyte_secret`
- [x] Unit & integration tests added and passing. Community members, please provide proof of success locally e.g: screenshot or copy-paste unit, integration, and acceptance test output. To run acceptance tests for a Python connector, follow instructions in the README. For java connectors run `./gradlew :airbyte-integrations:connectors:<name>:integrationTest`.
- [ ] Code reviews completed
- [x] Documentation updated
    - [ ] Connector's `README.md`
    - [ ] Connector's `bootstrap.md`. See [description and examples](https://docs.google.com/document/d/1ypdgmwmEHWv-TrO4_YOQ7pAJGVrMp5BOkEVh831N260/edit?usp=sharing)
    - [x] Changelog updated in `docs/integrations/<source or destination>/<name>.md` including changelog. See changelog [example](https://docs.airbyte.io/integrations/sources/stripe#changelog)
- [x] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)

### Airbyter

If this is a community PR, the Airbyte engineer reviewing this PR is responsible for the below items.

- [ ] Create a non-forked branch based on this PR and test the below items on it
- [ ] Build is successful
- [ ] If new credentials are required for use in CI, add them to GSM. [Instructions](https://docs.airbyte.io/connector-development#using-credentials-in-ci).
- [ ] [`/test connector=connectors/<name>` command](https://docs.airbyte.io/connector-development#updating-an-existing-connector) is passing
- [ ] New Connector version released on Dockerhub and connector version bumped by running the `/publish` command described [here](https://docs.airbyte.io/connector-development#updating-an-existing-connector)

</details>

<details><summary><strong>Connector Generator</strong></summary>

- [ ] Issue acceptance criteria met
- [ ] PR name follows [PR naming conventions](https://docs.airbyte.com/contributing-to-airbyte/issues-and-pull-requests)
- [ ] If adding a new generator, add it to the [list of scaffold modules being tested](https://github.com/airbytehq/airbyte/blob/master/airbyte-integrations/connector-templates/generator/build.gradle#L41)
- [ ] The generator test modules (all connectors with `-scaffold` in their name) have been updated with the latest scaffold by running `./gradlew :airbyte-integrations:connector-templates:generator:testScaffoldTemplates` then checking in your changes
- [ ] Documentation which references the generator is updated as needed

</details>

## Tests

<details><summary><strong>Unit</strong></summary>

*Put your unit tests output here.*

</details>

<details><summary><strong>Integration</strong></summary>

*Put your integration tests output here.*

</details>

<details><summary><strong>Acceptance</strong></summary>

*Put your acceptance tests output here.*

</details>

sajarin commented:
Hey @henriblancke thanks for the PR. This PR will be reviewed by a member of the community so we appreciate your patience as a maintainer is assigned to review the PR!

haithem-souala commented:
Hey @henriblancke, ill check this PR, so please, bear with me.

haithem-souala reviewed (changes requested):
Hey @henriblancke 

1 - Can you change the **documentationUrl** in `spec.json` to https://docs.airbyte.com/integrations/sources/retently

2- `companies` stream is not allowing more than 100 records, so with 1000 as limit, the validation fail, can you check that.

3- Can you add this PR url in the `retently.md` just like the older version

henriblancke commented:
@haithem-souala thanks for the review!

1) Added the `documentationUrl` in `specs.json`
2) I can't find anything explicitly in the retently documentation that states that the `companies` endpoint is different than the other endpoints. All the other endpoints accept `limit: 1000`. We don't use `companies` in our retently instance so I can't validate that it can accept `limit: 1000`.
3) Updated, added the PR url

haithem-souala commented:
> @haithem-souala thanks for the review!
> 
> 1. Added the `documentationUrl` in `specs.json`
> 2. I can't find anything in the retently documentation explicitly stating that the `companies` endpoint is different than the other endpoints and can only accept a limit of up to 100. All the other endpoints accept `limit: 1000`. We don't use `companies` in our retently instance so I can't validate that it can accept `limit: 1000`.
> 3. Updated, added the PR url

Thanx,
About the `companies`  API, i cannot find either anything about it in the documentation, but if you test using curl or postman, you will get a 400, and even it's not used in your instance, it should work for other users.

![Capture d’écran 2022-11-18 à 16 53 54](https://user-images.githubusercontent.com/75131226/202748280-b5dc40ee-8cca-4c40-9804-6eb27702ddfd.png)

haithem-souala commented on airbyte-integrations/connectors/source-retently/setup.py:
```diff
@@ -6,7 +6,7 @@
 
 MAIN_REQUIREMENTS = [
-    "airbyte-cdk",
+    "airbyte-cdk~=0.2",
```
Can you please explain this change?

henriblancke replied:
No direct explanation, except for trying to pin it to a specific version like most other connectors I think? I can remove the dependency pin if you prefer?

henriblancke commented:
> > @haithem-souala thanks for the review!
> > 
> > 1. Added the `documentationUrl` in `specs.json`
> > 2. I can't find anything in the retently documentation explicitly stating that the `companies` endpoint is different than the other endpoints and can only accept a limit of up to 100. All the other endpoints accept `limit: 1000`. We don't use `companies` in our retently instance so I can't validate that it can accept `limit: 1000`.
> > 3. Updated, added the PR url
> 
> Thank you! About the `companies` API, i cannot find either anything about it in the documentation, but if you test using curl or postman, you will get a 400, and even if it's not used in your instance, it should work for other users.
> 
> I suggest to set the limit to 100 for `companies` stream.
> 
> ![Capture d’écran 2022-11-18 à 16 53 54](https://user-images.githubusercontent.com/75131226/202748280-b5dc40ee-8cca-4c40-9804-6eb27702ddfd.png)

@haithem-souala I've updated the limit to be 100 for companies.

haithem-souala commented:
Last comment and you good to go, you need to declare all streams in` configured_catalog.json`

henriblancke commented:
> Last comment and you good to go, you need to declare all streams in`configured_catalog.json`, feedback, campaigns, outbox and templates are missing

@haithem-souala 👍 done

haithem-souala commented:
> > Last comment and you good to go, you need to declare all streams in`configured_catalog.json`, feedback, campaigns, outbox and templates are missing
> 
> @haithem-souala 👍 done

You missed the one in sample_files/configured_catalog.json

haithem-souala commented:
hey @sajarin, changes looks solide, can you run the /test command please!

henriblancke commented:
> > > Last comment and you good to go, you need to declare all streams in`configured_catalog.json`, feedback, campaigns, outbox and templates are missing
> > 
> > 
> > @haithem-souala 👍 done
> 
> Hey @henriblancke, You forgot sample_files/configured_catalog.json, there are 2 configured_catalog.json files

@haithem-souala oops missed that one, should be updated now

haithem-souala commented:
Hey @sajarin

I think this PR is ready, can you run the /test please?

sajarin commented:
/test connector=connectors/source-retently

marcosmarxm commented:
/test connector=connectors/source-retently

haithem-souala commented on airbyte-integrations/connectors/source-retently/unit_tests/test_streams.py:
```diff
@@ -18,10 +18,16 @@ def patch_base_class(mocker):
+    expected_params = {'limit': 100}
+    assert stream.request_params(**inputs) == expected_params
+
+def test_request_params_other(patch_base_class):
```
```suggestion

def test_request_params_other(patch_base_class):
```

Tests are not passing because of the blank line missed.

haithem-souala commented:
@marcosmarxm, @sajarin could you run the /test please?

marcosmarxm commented:
Hello :wave:, first thank you for this amazing contribution. 

We really appreciate the effort you've made to improve the project.
We ask you patience for the code review. Last month our team was focused on Hacktoberfest event and that probably left some PR without the proper feedback. And this week, due to the Thanksgiving US Holiday, most our team is out of office with their families. Another important piece of information why code won't be merge this week is: as a safety measure the core team has decided to freeze merging code to main branch to keep the release stable. Next week we'll return to you with the proper code review and update the status of your contribution.

If you have any questions feel free to send me a message in Slack!
Thanks!

sajarin commented:
/test connector=connectors/source-retently

haithem-souala commented:
@sajarin test command seems broken!

sajarin commented:
/test connector=connectors/source-retently

sajarin commented:
/test connector=connectors/source-retently

sajarin commented:
/test connector=connectors/source-retently
